stats/opentelemetry/opentelemetry.go

```diff
@@ -0,0 +1,239 @@
+func getCallInfo(ctx context.Context) *callInfo {
+	ci, _ := ctx.Value(callInfoKey{}).(*callInfo)
+	return ci
+}
+
+// rpcInfo is RPC information scoped to the RPC attempt life span client side,
+// and the RPC life span server side.
+type rpcInfo struct {
+	mi *metricsInfo
+}
+
+type rpcInfoKey struct{}
+
+func setRPCInfo(ctx context.Context, ri *rpcInfo) context.Context {
+	return context.WithValue(ctx, rpcInfoKey{}, ri)
+}
+
+// getRPCInfo returns the rpcInfo stored in the context, or nil
+// if there isn't one.
+func getRPCInfo(ctx context.Context) *rpcInfo {
+	ri, _ := ctx.Value(rpcInfoKey{}).(*rpcInfo)
+	return ri
+}
+
+func removeLeadingSlash(mn string) string {
+	return strings.TrimLeft(mn, "/")
+}
+
+// metricsInfo is RPC information scoped to the RPC attempt life span client
+// side, and the RPC life span server side.
+type metricsInfo struct {
+	// access these counts atomically for hedging in the future:
+	// number of bytes after compression (within each message) from side (client
+	// || server).
+	sentCompressedBytes int64
+	// number of compressed bytes received (within each message) received on
+	// side (client || server).
+	recvCompressedBytes int64
+
+	startTime time.Time
+	method    string
+	authority string
+}
+
+// registeredMetrics are the set of metrics to record with for the OpenTelemetry
+// component.
+type registeredMetrics struct {
+	// "grpc.client.attempt.started"
+	clientAttemptStarted metric.Int64Counter
+	// "grpc.client.attempt.duration"
+	clientAttemptDuration metric.Float64Histogram
+	// "grpc.client.attempt.sent_total_compressed_message_size"
+	clientAttemptSentTotalCompressedMessageSize metric.Int64Histogram
+	// "grpc.client.attempt.rcvd_total_compressed_message_size"
+	clientAttemptRcvdTotalCompressedMessageSize metric.Int64Histogram
+
+	// per call client metrics:
+	clientCallDuration metric.Float64Histogram
+
+	// "grpc.server.call.started"
+	serverCallStarted metric.Int64Counter
+	// "grpc.server.call.sent_total_compressed_message_size"
+	serverCallSentTotalCompressedMessageSize metric.Int64Histogram
+	// "grpc.server.call.rcvd_total_compressed_message_size"
+	serverCallRcvdTotalCompressedMessageSize metric.Int64Histogram
+	// "grpc.server.call.duration"
+	serverCallDuration metric.Float64Histogram
+}
+
+// Users of this component should use these bucket boundaries as part of their
+// SDK MeterProvider passed in. This component sends this as "advice" to the
+// API, which works, however this stability is not guaranteed, so for safety the
+// SDK Meter Provider should set these bounds.
+var (
+	// DefaultLatencyBounds are the default bounds for latency metrics.
+	DefaultLatencyBounds = []float64{0, 0.00001, 0.00005, 0.0001, 0.0003, 0.0006, 0.0008, 0.001, 0.002, 0.003, 0.004, 0.005, 0.006, 0.008, 0.01, 0.013, 0.016, 0.02, 0.025, 0.03, 0.04, 0.05, 0.065, 0.08, 0.1, 0.13, 0.16, 0.2, 0.25, 0.3, 0.4, 0.5, 0.65, 0.8, 1, 2, 5, 10, 20, 50, 100} // provide "advice" through API, SDK should set this too
```

Comment:
Do we need to export these?  Can we set them in this plugin somehow?

Reply:
Yes, we need to but only ever so slightly. The "advice" I create works, but is technically unstable, so the technically certain way to use this is for users to set it on their SDK passed in, which has no instability.

Reply:
"Unstable" how?